import grpc
import strawberry
from typing import List, Optional
from strawberry.fastapi import GraphQLRouter
from fastapi import FastAPI, HTTPException
import logging
from contextlib import asynccontextmanager

# –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ gRPC –º–æ–¥—É–ª–∏
import catalog_pb2
import catalog_pb2_grpc
import order_pb2
import order_pb2_grpc

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger('GraphQLGateway')

# –ö–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—è –∞–¥—Ä–µ—Å–æ–≤ –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–æ–≤
SERVICE_CONFIG = {
    'catalog': 'catalog-service:50052',
    'order': 'order-service:50054',
}

# –ì–ª–æ–±–∞–ª—å–Ω—ã–µ gRPC –∫–∞–Ω–∞–ª—ã –∏ —Å—Ç–∞–±—ã (–∫–ª–∏–µ–Ω—Ç—ã)
catalog_channel = None
order_channel = None
catalog_stub = None
order_stub = None

@asynccontextmanager
async def lifespan(app: FastAPI):
    """–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –∂–∏–∑–Ω–µ–Ω–Ω—ã–º —Ü–∏–∫–ª–æ–º –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è: —Å–æ–∑–¥–∞–Ω–∏–µ –∏ –∑–∞–∫—Ä—ã—Ç–∏–µ gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–π."""
    global catalog_channel, order_channel, catalog_stub, order_stub
    
    # –°–æ–∑–¥–∞–µ–º —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è –ø—Ä–∏ –∑–∞–ø—É—Å–∫–µ
    logger.info("–ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–π —Å –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–∞–º–∏...")
    try:
        catalog_channel = grpc.insecure_channel(SERVICE_CONFIG['catalog'])
        order_channel = grpc.insecure_channel(SERVICE_CONFIG['order'])
        
        # –°–æ–∑–¥–∞–µ–º –∫–ª–∏–µ–Ω—Ç–æ–≤ (—Å—Ç–∞–±—ã)
        catalog_stub = catalog_pb2_grpc.CatalogServiceStub(catalog_channel)
        order_stub = order_pb2_grpc.OrderServiceStub(order_channel)
        
        logger.info("‚úÖ gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω—ã")
        yield
    except Exception as e:
        logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ —É—Å—Ç–∞–Ω–æ–≤–∫–µ gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–π: {e}")
        raise
    finally:
        # –ó–∞–∫—Ä—ã–≤–∞–µ–º —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è –ø—Ä–∏ –æ—Å—Ç–∞–Ω–æ–≤–∫–µ
        logger.info("–ó–∞–∫—Ä—ã—Ç–∏–µ gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–π...")
        if catalog_channel:
            catalog_channel.close()
        if order_channel:
            order_channel.close()
        logger.info("gRPC —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è –∑–∞–∫—Ä—ã—Ç—ã")

# ==================== GraphQL –¢–ò–ü–´ ====================

@strawberry.type
class Product:
    id: str
    name: str
    description: Optional[str] = None
    price: float
    category_id: str
    quantity: int
    image_url: Optional[str] = None

@strawberry.type
class OrderItem:
    product_id: str
    product_name: str
    price: float
    quantity: int
    total: float

@strawberry.type
class Order:
    order_id: str
    user_id: str
    cart_id: Optional[str] = None
    items: List[OrderItem]
    subtotal: float
    tax: float
    shipping_cost: float
    total_amount: float
    status: str
    shipping_address: str
    payment_method: str
    created_at: str
    updated_at: str

@strawberry.type
class User:
    id: str
    email: str
    role: str

@strawberry.type
class OrderResult:
    success: bool
    order_id: Optional[str] = None
    message: str

# ==================== GraphQL –ó–ê–ü–†–û–°–´ (Queries) ====================

@strawberry.type
class Query:
    
    @strawberry.field
    async def product(self, id: str) -> Optional[Product]:
        """–ü–æ–ª—É—á–∏—Ç—å –ø—Ä–æ–¥—É–∫—Ç –ø–æ ID"""
        try:
            logger.info(f"–ó–∞–ø—Ä–æ—Å –ø—Ä–æ–¥—É–∫—Ç–∞ —Å ID: {id}")
            response = catalog_stub.GetProduct(catalog_pb2.GetProductRequest(product_id=id))
            
            return Product(
                id=response.id,
                name=response.name,
                description=response.description,
                price=response.price,
                category_id=response.category_id,
                quantity=response.quantity,
                image_url=response.image_url
            )
        except grpc.RpcError as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –ø—Ä–æ–¥—É–∫—Ç–∞ {id}: {e.details()}")
            if e.code() == grpc.StatusCode.NOT_FOUND:
                return None
            raise HTTPException(status_code=500, detail=f"Catalog service error: {e.details()}")
    
    @strawberry.field
    async def products(
        self, 
        category_id: Optional[str] = None,
        page: Optional[int] = 1,
        limit: Optional[int] = 10
    ) -> List[Product]:
        """–ü–æ–ª—É—á–∏—Ç—å —Å–ø–∏—Å–æ–∫ –ø—Ä–æ–¥—É–∫—Ç–æ–≤ —Å —Ñ–∏–ª—å—Ç—Ä–∞—Ü–∏–µ–π –∏ –ø–∞–≥–∏–Ω–∞—Ü–∏–µ–π"""
        try:
            logger.info(f"–ó–∞–ø—Ä–æ—Å —Å–ø–∏—Å–∫–∞ –ø—Ä–æ–¥—É–∫—Ç–æ–≤. –ö–∞—Ç–µ–≥–æ—Ä–∏—è: {category_id}, –°—Ç—Ä–∞–Ω–∏—Ü–∞: {page}, –õ–∏–º–∏—Ç: {limit}")
            
            response = catalog_stub.ListProducts(catalog_pb2.ListProductsRequest(
                category_id=category_id or "",
                page=page,
                limit=limit
            ))
            
            products_list = []
            for product in response.products:
                products_list.append(Product(
                    id=product.id,
                    name=product.name,
                    description=product.description,
                    price=product.price,
                    category_id=product.category_id,
                    quantity=product.quantity,
                    image_url=product.image_url
                ))
            
            logger.info(f"–ü–æ–ª—É—á–µ–Ω–æ {len(products_list)} –ø—Ä–æ–¥—É–∫—Ç–æ–≤")
            return products_list
            
        except grpc.RpcError as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Å–ø–∏—Å–∫–∞ –ø—Ä–æ–¥—É–∫—Ç–æ–≤: {e.details()}")
            raise HTTPException(status_code=500, detail=f"Catalog service error: {e.details()}")
    
    @strawberry.field
    async def orders(self, user_id: str) -> List[Order]:
        """–ü–æ–ª—É—á–∏—Ç—å –≤—Å–µ –∑–∞–∫–∞–∑—ã –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è"""
        try:
            logger.info(f"–ó–∞–ø—Ä–æ—Å –∑–∞–∫–∞–∑–æ–≤ –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è: {user_id}")
            
            response = order_stub.ListOrders(order_pb2.ListOrdersRequest(
                user_id=user_id,
                page=1,
                limit=100  # –î–æ—Å—Ç–∞—Ç–æ—á–Ω–æ –±–æ–ª—å—à–æ–µ —á–∏—Å–ª–æ, —á—Ç–æ–±—ã –ø–æ–ª—É—á–∏—Ç—å –≤—Å–µ –∑–∞–∫–∞–∑—ã
            ))
            
            orders_list = []
            for order in response.orders:
                # –ü—Ä–µ–æ–±—Ä–∞–∑—É–µ–º —ç–ª–µ–º–µ–Ω—Ç—ã –∑–∞–∫–∞–∑–∞
                order_items = []
                for item in order.items:
                    order_items.append(OrderItem(
                        product_id=item.product_id,
                        product_name=item.product_name,
                        price=item.price,
                        quantity=item.quantity,
                        total=item.total
                    ))
                
                orders_list.append(Order(
                    order_id=order.order_id,
                    user_id=order.user_id,
                    cart_id=order.cart_id,
                    items=order_items,
                    subtotal=order.subtotal,
                    tax=order.tax,
                    shipping_cost=order.shipping_cost,
                    total_amount=order.total_amount,
                    status=order.status,
                    shipping_address=order.shipping_address,
                    payment_method=order.payment_method,
                    created_at=order.created_at,
                    updated_at=order.updated_at
                ))
            
            logger.info(f"–ü–æ–ª—É—á–µ–Ω–æ {len(orders_list)} –∑–∞–∫–∞–∑–æ–≤ –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è {user_id}")
            return orders_list
            
        except grpc.RpcError as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –∑–∞–∫–∞–∑–æ–≤: {e.details()}")
            if e.code() == grpc.StatusCode.NOT_FOUND:
                return []  # –ï—Å–ª–∏ –∑–∞–∫–∞–∑–æ–≤ –Ω–µ—Ç, –≤–æ–∑–≤—Ä–∞—â–∞–µ–º –ø—É—Å—Ç–æ–π —Å–ø–∏—Å–æ–∫
            raise HTTPException(status_code=500, detail=f"Order service error: {e.details()}")
    
    @strawberry.field
    async def user_orders_with_products(self, user_id: str) -> List[Order]:
        """
        –ö–æ–º–±–∏–Ω–∏—Ä–æ–≤–∞–Ω–Ω—ã–π –∑–∞–ø—Ä–æ—Å: –∑–∞–∫–∞–∑—ã –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è —Å –ø–æ–ª–Ω–æ–π –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–µ–π –æ –ø—Ä–æ–¥—É–∫—Ç–∞—Ö.
        –î–µ–º–æ–Ω—Å—Ç—Ä–∏—Ä—É–µ—Ç –º–æ—â—å GraphQL –¥–ª—è –∞–≥—Ä–µ–≥–∞—Ü–∏–∏ –¥–∞–Ω–Ω—ã—Ö.
        """
        try:
            logger.info(f"–ö–æ–º–±–∏–Ω–∏—Ä–æ–≤–∞–Ω–Ω—ã–π –∑–∞–ø—Ä–æ—Å –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è: {user_id}")
            
            # 1. –ü–æ–ª—É—á–∞–µ–º –∑–∞–∫–∞–∑—ã –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
            orders_response = order_stub.ListOrders(order_pb2.ListOrdersRequest(
                user_id=user_id,
                page=1,
                limit=50
            ))
            
            orders_with_details = []
            
            # 2. –î–ª—è –∫–∞–∂–¥–æ–≥–æ –∑–∞–∫–∞–∑–∞ –æ–±–æ–≥–∞—â–∞–µ–º –¥–∞–Ω–Ω—ã–µ –æ –ø—Ä–æ–¥—É–∫—Ç–∞—Ö
            for order in orders_response.orders:
                enriched_items = []
                
                for item in order.items:
                    # –ó–∞–ø—Ä–∞—à–∏–≤–∞–µ–º –¥–µ—Ç–∞–ª—å–Ω—É—é –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –ø—Ä–æ–¥—É–∫—Ç–µ –∏–∑ –∫–∞—Ç–∞–ª–æ–≥–∞
                    try:
                        product_detail = catalog_stub.GetProduct(
                            catalog_pb2.GetProductRequest(product_id=item.product_id)
                        )
                        
                        # –°–æ–∑–¥–∞–µ–º –æ–±–æ–≥–∞—â–µ–Ω–Ω—ã–π —ç–ª–µ–º–µ–Ω—Ç –∑–∞–∫–∞–∑–∞
                        enriched_items.append(OrderItem(
                            product_id=item.product_id,
                            product_name=product_detail.name,  # –ë–µ—Ä–µ–º –∏–º—è –∏–∑ –∫–∞—Ç–∞–ª–æ–≥–∞
                            price=item.price,
                            quantity=item.quantity,
                            total=item.total
                        ))
                    except grpc.RpcError:
                        # –ï—Å–ª–∏ –Ω–µ —É–¥–∞–ª–æ—Å—å –ø–æ–ª—É—á–∏—Ç—å –¥–µ—Ç–∞–ª–∏ –ø—Ä–æ–¥—É–∫—Ç–∞, –∏—Å–ø–æ–ª—å–∑—É–µ–º –±–∞–∑–æ–≤—É—é –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é
                        enriched_items.append(OrderItem(
                            product_id=item.product_id,
                            product_name=item.product_name,
                            price=item.price,
                            quantity=item.quantity,
                            total=item.total
                        ))
                
                # –°–æ–∑–¥–∞–µ–º –æ–±–æ–≥–∞—â–µ–Ω–Ω—ã–π –∑–∞–∫–∞–∑
                orders_with_details.append(Order(
                    order_id=order.order_id,
                    user_id=order.user_id,
                    cart_id=order.cart_id,
                    items=enriched_items,
                    subtotal=order.subtotal,
                    tax=order.tax,
                    shipping_cost=order.shipping_cost,
                    total_amount=order.total_amount,
                    status=order.status,
                    shipping_address=order.shipping_address,
                    payment_method=order.payment_method,
                    created_at=order.created_at,
                    updated_at=order.updated_at
                ))
            
            logger.info(f"–°–æ–∑–¥–∞–Ω–æ {len(orders_with_details)} –æ–±–æ–≥–∞—â–µ–Ω–Ω—ã—Ö –∑–∞–∫–∞–∑–æ–≤")
            return orders_with_details
            
        except grpc.RpcError as e:
            logger.error(f"–û—à–∏–±–∫–∞ –≤ –∫–æ–º–±–∏–Ω–∏—Ä–æ–≤–∞–Ω–Ω–æ–º –∑–∞–ø—Ä–æ—Å–µ: {e.details()}")
            raise HTTPException(status_code=500, detail=f"Service error: {e.details()}")
    
    @strawberry.field
    async def health(self) -> str:
        """–ü—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è —Å–µ—Ä–≤–∏—Å–∞"""
        try:
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º –¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∫–∞—Ç–∞–ª–æ–≥–∞
            catalog_stub.ListProducts(catalog_pb2.ListProductsRequest(limit=1))
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º –¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–∫–∞–∑–æ–≤
            order_stub.ListOrders(order_pb2.ListOrdersRequest(user_id="test", limit=1))
            
            return "‚úÖ –í—Å–µ —Å–∏—Å—Ç–µ–º—ã —Ä–∞–±–æ—Ç–∞—é—Ç –Ω–æ—Ä–º–∞–ª—å–Ω–æ"
        except grpc.RpcError as e:
            logger.warning(f"–ü—Ä–æ–±–ª–µ–º—ã —Å –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ–º –∫ —Å–µ—Ä–≤–∏—Å–∞–º: {e.details()}")
            return f"‚ö†Ô∏è –ü—Ä–æ–±–ª–µ–º—ã —Å –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ–º: {e.details()}"
        except Exception as e:
            logger.error(f"–ù–µ–ø—Ä–µ–¥–≤–∏–¥–µ–Ω–Ω–∞—è –æ—à–∏–±–∫–∞ health check: {e}")
            return f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {str(e)}"

# ==================== GraphQL –ú–£–¢–ê–¶–ò–ò (Mutations) ====================

@strawberry.type
class Mutation:
    
    @strawberry.mutation
    async def create_order(
        self, 
        user_id: str,
        product_id: str,
        quantity: int,
        shipping_address: str,
        payment_method: str = "CREDIT_CARD"
    ) -> OrderResult:
        """–°–æ–∑–¥–∞—Ç—å –Ω–æ–≤—ã–π –∑–∞–∫–∞–∑"""
        try:
            logger.info(f"–°–æ–∑–¥–∞–Ω–∏–µ –∑–∞–∫–∞–∑–∞ –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è {user_id}, —Ç–æ–≤–∞—Ä: {product_id}, –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ: {quantity}")
            
            # 1. –ü–æ–ª—É—á–∞–µ–º –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –ø—Ä–æ–¥—É–∫—Ç–µ –∏–∑ –∫–∞—Ç–∞–ª–æ–≥–∞
            product_response = catalog_stub.GetProduct(
                catalog_pb2.GetProductRequest(product_id=product_id)
            )
            
            # 2. –ü–æ–¥–≥–æ—Ç–∞–≤–ª–∏–≤–∞–µ–º —ç–ª–µ–º–µ–Ω—Ç –∑–∞–∫–∞–∑–∞
            order_item = order_pb2.OrderItem(
                product_id=product_id,
                product_name=product_response.name,
                price=product_response.price,
                quantity=quantity,
                total=product_response.price * quantity
            )
            
            # 3. –°–æ–∑–¥–∞–µ–º –∑–∞–∫–∞–∑
            order_response = order_stub.CreateOrder(order_pb2.CreateOrderRequest(
                user_id=user_id,
                cart_id=f"cart_{user_id}",  # –í —Ä–µ–∞–ª—å–Ω–æ–º –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–∏ cart_id –Ω—É–∂–Ω–æ –ø–æ–ª—É—á–∞—Ç—å –∏–∑ –∫–æ—Ä–∑–∏–Ω—ã
                shipping_address=shipping_address,
                payment_method=payment_method,
                items=[order_item]
            ))
            
            logger.info(f"–ó–∞–∫–∞–∑ —Å–æ–∑–¥–∞–Ω —É—Å–ø–µ—à–Ω–æ: {order_response.order_id}")
            
            return OrderResult(
                success=True,
                order_id=order_response.order_id,
                message=f"–ó–∞–∫–∞–∑ {order_response.order_id} —É—Å–ø–µ—à–Ω–æ —Å–æ–∑–¥–∞–Ω"
            )
            
        except grpc.RpcError as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ —Å–æ–∑–¥–∞–Ω–∏–∏ –∑–∞–∫–∞–∑–∞: {e.details()}")
            return OrderResult(
                success=False,
                message=f"–û—à–∏–±–∫–∞ –ø—Ä–∏ —Å–æ–∑–¥–∞–Ω–∏–∏ –∑–∞–∫–∞–∑–∞: {e.details()}"
            )

# ==================== –ù–ê–°–¢–†–û–ô–ö–ê –ò –ó–ê–ü–£–°–ö ====================

# –°–æ–∑–¥–∞–µ–º GraphQL —Å—Ö–µ–º—É
schema = strawberry.Schema(query=Query, mutation=Mutation)

# –°–æ–∑–¥–∞–µ–º FastAPI –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–µ
graphql_app = GraphQLRouter(schema, path="/graphql")

app = FastAPI(
    title="GraphQL Gateway –¥–ª—è –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–æ–≤ –º–∞–≥–∞–∑–∏–Ω–∞",
    description="–ê–≥—Ä–µ–≥–∏—Ä—É–µ—Ç –¥–∞–Ω–Ω—ã–µ –∏–∑ Catalog –∏ Order —Å–µ—Ä–≤–∏—Å–æ–≤",
    version="1.0.0",
    lifespan=lifespan
)

# –ü–æ–¥–∫–ª—é—á–∞–µ–º GraphQL —Ä–æ—É—Ç–µ—Ä
app.include_router(graphql_app)

# –î–æ–±–∞–≤–ª—è–µ–º REST —ç–Ω–¥–ø–æ–∏–Ω—Ç—ã –¥–ª—è –æ–±—Ä–∞—Ç–Ω–æ–π —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏
@app.get("/")
async def root():
    """–ö–æ—Ä–Ω–µ–≤–æ–π —ç–Ω–¥–ø–æ–∏–Ω—Ç —Å –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–µ–π –æ —Å–µ—Ä–≤–∏—Å–µ"""
    return {
        "service": "GraphQL Gateway",
        "version": "1.0.0",
        "description": "–ê–≥—Ä–µ–≥–∞—Ç–æ—Ä –¥–∞–Ω–Ω—ã—Ö –¥–ª—è –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–æ–≤ –º–∞–≥–∞–∑–∏–Ω–∞",
        "endpoints": {
            "graphql": "/graphql",
            "health": "/health",
            "openapi": "/docs"
        },
        "connected_services": list(SERVICE_CONFIG.keys())
    }

@app.get("/health")
async def health_check():
    """REST —ç–Ω–¥–ø–æ–∏–Ω—Ç –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –∑–¥–æ—Ä–æ–≤—å—è"""
    try:
        # –ë—ã—Å—Ç—Ä–∞—è –ø—Ä–æ–≤–µ—Ä–∫–∞ —Å–µ—Ä–≤–∏—Å–æ–≤
        catalog_stub.ListProducts(catalog_pb2.ListProductsRequest(limit=1))
        return {"status": "healthy", "service": "GraphQL Gateway"}
    except Exception as e:
        return {"status": "unhealthy", "error": str(e)}, 503

@app.get("/api/products")
async def rest_products(category_id: str = None, page: int = 1, limit: int = 10):
    """REST —ç–Ω–¥–ø–æ–∏–Ω—Ç –¥–ª—è –ø–æ–ª—É—á–µ–Ω–∏—è –ø—Ä–æ–¥—É–∫—Ç–æ–≤ (–¥–ª—è –æ–±—Ä–∞—Ç–Ω–æ–π —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏)"""
    products = await Query().products(category_id, page, limit)
    return {"products": products, "total": len(products)}

# –¢–æ—á–∫–∞ –≤—Ö–æ–¥–∞ –¥–ª—è –∑–∞–ø—É—Å–∫–∞
if __name__ == "__main__":
    import uvicorn
    logger.info("=" * 60)
    logger.info("üöÄ –ó–∞–ø—É—Å–∫ GraphQL Gateway –Ω–∞ Strawberry...")
    logger.info(f"üì° –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–∏—Å–∞–º: {SERVICE_CONFIG}")
    logger.info("üìç GraphQL Playground: http://localhost:8081/graphql")
    logger.info("üìç REST Health Check: http://localhost:8081/health")
    logger.info("=" * 60)
    
    uvicorn.run(
        app, 
        host="0.0.0.0", 
        port=8081,
        log_level="info"
    )